2nd revocation proposal (#205)
smart-on-fhir/health-cards: docs/index.md

issuer's JWK set in its `jwks.json` file. Retired private keys that are no longer used to sign Health Cards SHALL be destroyed.
Older public key entries that are needed to validate previously
signed Health Cards SHALL remain in the JWK set for as long as the corresponding Health Cards
are clinically relevant. However, if a private signing key is compromised, then the issuer SHALL immediately remove the corresponding public key from the JWK set in its `jwks.json` file and request revocation of all X.509 certificates bound to that public key; verifiers will from then on reject all Health Cards signed using that key.

### Revocation

> Individual Health Cards MAY be revoked using a revocation identifier property `rid` encoded in the `vc` claim of the JWT. This should be a short identifier, meaningless to the verifiers; the only constraint is that the identifier SHALL use the base64url alphabet (but doesn’t need to be base64url encoded). Issuers MAY use application-specific user identifiers for this purpose, but since these could be publicly listed in revocation lists, issuers SHOULD use a one-way transformation of the data combined with enough entropy to prevent reversal. It is RECOMMENDED to use the base64url encoding of the first 64 bits of the output of HMAC-SHA-256 (as specified in [RFC 4868](https://tools.ietf.org/html/rfc4868)) using a 256-bit random secret key concatenated with the `<<kid>>`.

Reviewer: Can we impose an absolute upper bound of how long a `rid` can be?

Author: > Can we impose an absolute upper bound of how long a `rid` can be?

There is a natural goal of keeping these small to 1) still fit in v22 QR, and 2) keep the CRLs small, but we could define an upper bound if we need to enforce this. What would you suggest?

Reviewer: Yeah, agree with the implicit goal. So I presume it only makes sense to have an absolute upper limit. I'm not sure what a good value is, should we aim to allow 256 bits, even though we recommend taking the first 64 only? So, in base64, that's like 43 chars, can we limit to 50 chars? Too much? Too little?

Reviewer: OTOH should we just specify _how_ to calculate the `rid` and be done with it? I'm not sure what the risk of conflict is if we just take the first 64 bits.

Author: 64 bit is safe to avoid accidental collisions. The only "attacker" in this case would be the issuer itself, who could try to generate collisions, a pointless exercice. Since this does not need to be processed by wallets and verifiers, I don't think we need/should enforce an algorithm; but the recommended hmac approach is a good way to transform an ID meaningful to the issuer into an opaque one to everyone else.

Reviewer: Ok, but if we think 64 bits are enough, which is like ~11 base64 chars, then we can set a much lower limit?

To enable per-card revocation, the issuer creates, for each of its keys, a JSON Card Revocation List (CRL) file with the following content:
```json
{
"kid": "<<kid>>",
"method": "rid",
"ctr": "<<ctr>>",
"rids": [...]
}
```

where
- `"<<kid>>"` is the ID of the corresponding issuer key,
- `"rid"` identifies the revocation method specified in this framework; legacy cards can use different methods specified in external revocation profiles,
- `"<<ctr>>"` is counter indicating how many times this file has been updated; initial value is 1,

> -  `rids` is an array of revoked cards' identifiers `rid` (using the base64url alphabet) combined with optional timestamps.

Reviewer 2: ```suggestion
-  `rids` is an array of revoked cards' identifiers `rid` values. These values are represented as strings from the base64url alphabet, plus an optional timestamp suffix consisting of `.` followed by a numerical timestamp (e.g., `.1636977600`)
```

To revoke a Health Card issued under the key `"<<kid>>"`, an issuer adds its revocation identifier to the `rids` array of the corresponding `<<kid>>`'s revocation file. Since an issuer might want to invalidate a series of Health Cards associated with the user up to a certain time, the `rid` might be followed by a separator `.` a timestamp (encoded as the number of seconds from 1970-01-01T00:00:00Z UTC, as specified by RFC 7519). After updating the `rids` array (with one or more items), the `<<ctr>>` is incremented.

> As an example, the `rids` array `[AQPCj4wwk6Mt, lHKzqFUMjhs.1636977600]` marks as revoked any Health Cards with `rid` equal to `AQPCj4wwk6Mt` and Health Cards with `rid` equal to `lHKzqFUMjhs` issued before November 15, 2021 12:00:00 PM GMT.

Reviewer 2: ```suggestion
As an example, the `rids` array `["AQPCj4wwk6Mt", "lHKzqFUMjhs.1636977600"]` marks as revoked any Health Cards with `rid` equal to `AQPCj4wwk6Mt` and Health Cards with `rid` equal to `lHKzqFUMjhs` issued before November 15, 2021 12:00:00 PM GMT.
```

The per-key revocation file is made available at `https://"<<Issuer URL>>"/.well-known/crl/kid/"<<kid>>".json`, where
- `"<<Issuer URL>>"` is the issuer URL listed in the Health Card,
- `"<<kid>>"` is the key ID with which the Health Card was signed.

Issuers supporting this revocation method SHALL include in their published JWK set, for each key, a `crl-version` field encoding the update counter "<<ctr>>" for the corresponding revocation file.

> If the `crl-version` is present in the Issuer's JWK for key `<<kid>>`, Verifiers SHALL

Reviewer 2: Can we call this `crlVersion`? Would like to follow a common JSON naming schemes compatible with JS variable names...

Author: Done.

- Download the `https://"<<Issuer URL>>"/.well-known/crl/kid/"<<kid>>".json` file or use a cached version if the counter value has not changed since the last retrieval,

> - Reject the Health Card if the calculated `rid` is contained in the CRL's `rids` array and the Health Card’s `nbf` is value is before the timestamp, if present.

Reviewer 2: ```suggestion
- Reject the Health Card if the calculated `rid` is contained in the CRL's `rids` array and (if a timestamp suffix is present) the Health Card’s `nbf` is value is before the timestamp.
```
